practicas/fwgalde/P3/3_language_models_word_vectors.py

```diff
@@ -0,0 +1,462 @@
+
+
+ppmi_matrix = get_ppmi(joint_prob, marginal_prob)
+pmi = pd.DataFrame(data=ppmi_matrix, index=list(indices.keys()), columns=list(indices.keys()))
+pmi
+
+
+# ## ***Ejercicio 2:*** **Comparación de representaciones**
+
+# Aplica reducción de dimensionalidad (a 2D) de los vectores de la matríz con PPMI y de los vectores entrenados en español:
+# 
+# - Realiza un plot de 100 vectores aleatorios (que esten tanto en la matríz como en los vectores entrenados)
+# - Compara los resultados de los plots:
+#     - ¿Qué representación dirías que captura mejor relaciones semánticas?
+#     - Realiza un cuadro comparativo de ambos métodos con ventajas/desventajas
+# 
+# ### 📁 [Carpeta con vectores](https://drive.google.com/drive/folders/1reor2FGsfOB6m3AvfCE16NOHltAFjuvz?usp=drive_link)
+
+# ### ***Bibliotecas***
+
+# In[31]:
+
+
+from gensim.models import word2vec
+from sklearn.decomposition import PCA, TruncatedSVD
+import matplotlib.pyplot as plt
+
+
+# ### ***Variables globales***
+
+# In[23]:
+
+
+from enum import Enum
+
+class Algorithms(Enum):
+    CBOW = 0
+    SKIP_GRAM = 1
+
+
+# ### ***Funciones auxiliares***
+
+# In[24]:
+
+
+def load_model(model_path: str):
+    try:
+        print(model_path)
+        return word2vec.Word2Vec.load(model_path)
+    except:
+        print(f"[WARN] Model not found in path {model_path}")
+        return None
+
+
+# In[32]:
+
+
+def train_model(sentences: list, model_name: str, vector_size: int, window=5, workers=2, algorithm = Algorithms.CBOW):
+    model_name_params = f"{model_name}-vs{vector_size}-w{window}-{algorithm.name}.model"
+    model_path = "./" + model_name_params
+    print(f"TRAINING: {model_path}")
+    if algorithm in [Algorithms.CBOW, Algorithms.SKIP_GRAM]:
+        model = word2vec.Word2Vec(
+            sentences,
+            vector_size=vector_size,
+            window=window,
+            workers=workers,
+            sg = algorithm.value,
+            seed=42,
+            )
+    else:
+        print("[ERROR] algorithm not implemented yet :p")
+        return
+    try:
+        model.save(model_path)
+    except:
+        print(f"[ERROR] Saving model at {model_path}")
+    return model
+
+
+# ### ***Entrenando modelo word2vec***
+
+# In[33]:
+
+
+model = train_model(corpus_processed, model_name="my_corpus", 
+                    vector_size=100, window=5, workers=4, 
+                    algorithm=Algorithms.SKIP_GRAM)
+
+
+# ### ***Comparación de resultados***
+
+# In[34]:
+
+
+common_words = set(indices.keys()).intersection(set(model.wv.key_to_index.keys()))
+common_words = list(common_words)
+print(f"Número de palabras comunes: {len(common_words)}")
+
+selected_words = np.random.choice(common_words, size=100, replace=False)
+
+
+# In[35]:
+
+
+word2vec_vectors = np.array([model.wv[word] for word in selected_words])
+
+
+# In[36]:
+
+
+svd = TruncatedSVD(n_components=2, random_state=29)
+reduced_matrix = svd.fit_transform(ppmi_matrix)
+
+svd = PCA(n_components=2, random_state=29)
```

Comment:
Este nombre no sería correcto. Conviene cambiarlo. Además, sobrescribe el la variable anterior y puede propiciar errores.

```suggestion
pca = PCA(n_components=2, random_state=29)
```